### PR TITLE
Code walkthrough

### LINKED ISSUE #2684: High-level code walkthrough
It would be helpful to provide a high-level walk-through of the code. This has been asked for in a couple forms in the past: #114 #811.

After #2661 we'll have some pretty decent documentation of the production hosting which explains the architecture of the system pretty well.

### PR DESCRIPTION
**Pull request by paulmelnikow**
Updated based on comments received in the Google Doc.

Close #2684

### CODE REVIEW

File: doc/code-walkthrough.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+1.  The frontend (about 7% of the code)
+    1. [`frontend`][frontend]
+2.  The badge renderer (which is available as an npm package)
+    1.  [`gh-badges`][gh-badges]
+3.  The base service classes (about 8% of the code, and probably the most important
+    code in the codebase)
+    1.  [`core/base-service`][base-service]
+4.  The server code and a few related odds and ends
+    1.  `core/server`
```
Comment (chris48s) [inline doc/code-walkthrough.md]:
```suggestion
    1.  [`core/server`][server]
```

File: doc/code-walkthrough.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+5.  Helper code for token pooling and persistence (used to avoid GitHub rate limiting)
+    1.  `core/token-pooling`
```
Comment (chris48s) [inline doc/code-walkthrough.md]:
```suggestion
    1.  [`core/token-pooling`][token-pooling]
```

File: doc/code-walkthrough.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+8.  The badge suggestion endpoint (Note: it's tested as if it’s a service.)
+    1.  [`services/suggest.js`][suggest]
+
+[frontend]: https://github.com/badges/shields/tree/master/frontend
+[gh-badges]: https://github.com/badges/shields/tree/master/gh-badges
+[base-service]: https://github.com/badges/shields/tree/master/core/base-service
+[server]: https://github.com/badges/shields/tree/master/core/server
+[token-pooling]: https://github.com/badges/shields/tree/master/core/token-pooling
+[services]: https://github.com/badges/shields/tree/master/services
+[suggest]: https://github.com/badges/shields/tree/master/services/suggest.js
```
Comment (calebcartwright) [inline doc/code-walkthrough.md]: Should this be the `suggest` directory? https://github.com/badges/shields/tree/master/services/suggest

**Comment by calebcartwright** [comment]
This looks good! I also need to read up on GH beta suggested changes feature... Did not know about that!

**Comment by paulmelnikow** [comment]
> would it look slightly more natural with a `*`/• bullet, or does that stop the consecutive numbering from working? If so, we can leave it..

Yea, I'm pretty sure GitHub markdown doesn't let you nest bulleted lists inside of a numbered list.

File: doc/code-walkthrough.md (hunk 3 above)
Comment (paulmelnikow) [reply doc/code-walkthrough.md]: Oops, it's just not moved there yet.

Comment (paulmelnikow) [reply doc/code-walkthrough.md]: I opened #2873. Will update this to point to the pre-move file path in the meantime.

**Pull request merged**